src/Nethermind/Nethermind.Consensus/Stateless/WitnessCapturingTrieStore.cs

```diff
@@ -0,0 +1,69 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Trie;
+using Nethermind.Trie.Pruning;
+
+namespace Nethermind.Consensus.Stateless;
+
+/// <summary>
+/// WitnessCapturingTrieStore works by reading and writing to the passed in keyValueStore first as if it is an archive node.
+/// If a node is missing, then it will try to find from the base store.
+/// On reset the base db provider is expected to clear any diff which causes this overlay trie store to no longer
+/// see overlayed keys.
+/// </summary>
+public class WitnessCapturingTrieStore(IKeyValueStoreWithBatching keyValueStore, IReadOnlyTrieStore baseStore) : ITrieStore
+{
+    private readonly INodeStorage _nodeStorage = new NodeStorage(keyValueStore);
+
+    private readonly ConcurrentDictionary<Hash256, byte[]> _rlpCollector = new();
+
+    public byte[][] TouchedNodesRlp => _rlpCollector.Values.ToArray();
+
+    public void Dispose()
+    {
+        baseStore.Dispose();
+    }
+
+    public TrieNode FindCachedOrUnknown(Hash256? address, in TreePath path, Hash256 hash)
+    {
+        TrieNode node = baseStore.FindCachedOrUnknown(address, in path, hash);
+        if (node.NodeType != NodeType.Unknown)
+            _rlpCollector.TryAdd(node.Keccak, node.FullRlp.Span.ToArray());
```

Comment:
If it is already an array you save on allocation:

```suggestion
            _rlpCollector.TryAdd(node.Keccak, node.FullRlp.ToArray()!);
```